File: windows/src/RealtimeTranslator.Core/Realtime/SessionSubtitlePipeline.cs
using System;
using System.Threading.Tasks;
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.OpenAI;

namespace RealtimeTranslator.Core.Realtime;

/// <summary>
/// 字幕パイプラインの帳簿。<see cref="RealtimeSubtitleProcessor"/> と
/// <see cref="AudioLossTracker"/>、subtitle sequence、loss epoch を束ねる。
/// <c>InterpretationSession</c> の <c>_sync</c> ロック契約を引き継ぎ、独自ロックは持たない。
/// processor 呼び出しは呼び出し側が <c>_sync</c> を保持する前提。
/// </summary>
internal sealed class SessionSubtitlePipeline
{
    private readonly object _sync;
    private readonly TimeProvider _timeProvider;
    private readonly RealtimeSubtitleProcessor _processor = new();
    private readonly AudioLossTracker _audioLossTracker = new();
    private readonly Func<RealtimeEventFeed?> _activeFeedProvider;
    private readonly Func<RealtimeEventFeed> _fallbackFeedProvider;
    private readonly Action<RealtimeSubtitleUpdate> _emitUpdate;
    private readonly Func<Action?> _afterFailedSourceFallbackProvider;
    private int? _handledLossEpoch;
    private long _subtitleSequence;

    public SessionSubtitlePipeline(
        object sync,
        TimeProvider timeProvider,
        Func<RealtimeEventFeed?> activeFeedProvider,
        Func<RealtimeEventFeed> fallbackFeedProvider,
        Action<RealtimeSubtitleUpdate> emitUpdate,
        Func<Action?> afterFailedSourceFallbackProvider)
    {
        ArgumentNullException.ThrowIfNull(sync);
        ArgumentNullException.ThrowIfNull(timeProvider);
        ArgumentNullException.ThrowIfNull(activeFeedProvider);
        ArgumentNullException.ThrowIfNull(fallbackFeedProvider);
        ArgumentNullException.ThrowIfNull(emitUpdate);
        ArgumentNullException.ThrowIfNull(afterFailedSourceFallbackProvider);
        _sync = sync;
        _timeProvider = timeProvider;
        _activeFeedProvider = activeFeedProvider;
        _fallbackFeedProvider = fallbackFeedProvider;
        _emitUpdate = emitUpdate;
        _afterFailedSourceFallbackProvider = afterFailedSourceFallbackProvider;
    }

    public AudioLossMetrics AudioLossMetrics => _audioLossTracker.Metrics;

    public LanguagePair? ActiveLanguagePair => _processor.ActiveLanguagePair;

    public bool HasSelectedTranslationTarget => _processor.HasSelectedTranslationTarget;

    public int CurrentSourceLength => _processor.CurrentSourceLength;

    public bool IsCurrentSegmentTainted => _processor.IsCurrentSegmentTainted;

    /// <summary>新しい epoch を開始し、欠落処理済み epoch をリセットする（_sync 下で呼ぶ）。</summary>
    public void BeginEpoch(int epoch, LanguagePair languagePair)
    {
        _processor.BeginEpoch(epoch, languagePair);
        _handledLossEpoch = null;
    }

    public void DeactivateLanguagePair() => _processor.DeactivateLanguagePair();

    public void ClearBoundaryCandidate() => _processor.ClearBoundaryCandidate();

    public void ResetRoutingForNextSegment() => _processor.ResetRoutingForNextSegment();

    public void ResetAudioLoss() => _audioLossTracker.Reset();

    public AudioLossObservation ObserveAudio(
        int generation,
        long sequence,
        int discardedMilliseconds,
        int queueWaitMilliseconds,
        long atMilliseconds) =>
        _audioLossTracker.Observe(
            generation,
            sequence,
            discardedMilliseconds,
            queueWaitMilliseconds,
            atMilliseconds);

    public RealtimeSubtitleUpdate MarkAudioLoss() =>
        _processor.MarkAudioLoss(_timeProvider.GetUtcNow());

    public RealtimeSubtitleUpdate? DiscardFailedSource(string? itemId, string? eventId) =>
        _processor.DiscardFailedSource(itemId, eventId);

    public RealtimeSubtitleUpdate DiscardUnconfirmed() => _processor.DiscardUnconfirmed();

    public RealtimeSubtitleUpdate? Tick() => _processor.Tick(_timeProvider.GetUtcNow());

    public RealtimeSubtitleProcessingResult? Process(RealtimeTranslationStreamEvent streamEvent) =>
        _processor.Process(streamEvent, _timeProvider.GetUtcNow());

    /// <summary>
    /// 完全な原文+訳文ペアが assembler に残っていれば idle 待ちを飛ばして確定する。
    /// 停止・再接続・致命エラーで epoch/buffer を捨てる直前に呼び、字幕記録の欠落を防ぐ。
    /// </summary>
    public void FlushPendingFinalizeIfNeeded()
    {
        var feed = _activeFeedProvider();
        if (feed is { DeliveryState.DidLoseEvents: true })
        {
            HandleEventLoss(feed);
            return;
        }
        if (feed is { DeliveryState.HasPendingSourceFailure: true })
        {
            DiscardFailedSourceIfNeeded(feed);
            return;
        }

        RealtimeSubtitleUpdate? pending;
        lock (_sync)
        {
            pending = _processor.IsCurrentSegmentTainted
                ? _processor.DiscardUnconfirmed()
                : _processor.Tick(
                    _timeProvider.GetUtcNow() + RealtimeSubtitleAssembler.IdleFinalizeInterval);
        }

        if (pending is { } update)
        {
            EmitSubtitleUpdate(update);
        }
    }

    public void DiscardFailedSourceIfNeeded(RealtimeEventFeed feed)
    {
        if (!feed.DeliveryState.HasPendingSourceFailure)
        {
            return;
        }

        RealtimeSubtitleUpdate? invalidation;
        lock (_sync)
        {
            invalidation = _processor.DiscardFailedSource(null, null);
        }

        if (invalidation is { } failedUpdate)
        {
            EmitSubtitleUpdate(failedUpdate);
        }

        while (feed.DeliveryState.HasPendingSourceFailure)
        {
            feed.DeliveryState.NoteSourceFailureConsumed();
        }
        _afterFailedSourceFallbackProvider()?.Invoke();
    }

    /// <summary>
    /// 正常停止の close drain で channel に残った字幕イベントを assembler へ取り込む。
    /// session consumer は世代更新で既に止まっている前提。
    /// </summary>
    public async Task IngestStopDrainEventsAsync()
    {
        var feed = _activeFeedProvider();
        if (feed is { DeliveryState.DidLoseEvents: true })
        {
            HandleEventLoss(feed);
            return;
        }

        var events = feed?.Events ?? _fallbackFeedProvider().Events;
        while (await events.WaitToReadAsync().ConfigureAwait(false))
        {
            IngestAlreadyQueuedEvents();
        }
    }

    /// <summary>
    /// すでに channel にあるイベントだけを取り込む。WaitToRead しないので、
    /// ForceClose 前の再接続 teardown から呼んでも開いたままの channel で止まらない。
    /// </summary>
    public void IngestAlreadyQueuedEvents()
    {
        var feed = _activeFeedProvider();
        if (feed is { DeliveryState.DidLoseEvents: true })
        {
            HandleEventLoss(feed);
            return;
        }

        var currentFeed = feed ?? _fallbackFeedProvider();
        var events = currentFeed.Events;
        while (events.TryRead(out var streamEvent))
        {
            if (streamEvent.Event is RealtimeTranslationServerEvent.ServerError)
            {
                continue;
            }

            if (streamEvent.Event is RealtimeTranslationServerEvent.InputTranscriptFailed failed)
            {
                if (streamEvent.Epoch != currentFeed.Epoch)
                {
                    continue;
                }

                RealtimeSubtitleUpdate? invalidation;
                lock (_sync)
                {
                    invalidation = _processor.DiscardFailedSource(failed.ItemId, failed.EventId);
                }
                currentFeed.DeliveryState.NoteSourceFailureConsumed();

                if (invalidation is { } failedUpdate)
                {
                    EmitSubtitleUpdate(failedUpdate);
                }

                continue;
            }

            RealtimeSubtitleProcessingResult? result;
            lock (_sync)
            {
                if (feed is { DeliveryState.DidLoseEvents: true })
                {
                    result = null;
                }
                else
                {
                    result = _processor.Process(streamEvent, _timeProvider.GetUtcNow());
                }
            }

            if (result is { } processed)
            {
                foreach (var update in processed.Updates)
                {
                    EmitSubtitleUpdate(update);
                }
            }
        }
    }

    public bool HandleEventLoss(RealtimeEventFeed feed)
    {
        RealtimeSubtitleUpdate? invalidation = null;
        lock (_sync)
        {
            if (!feed.DeliveryState.DidLoseEvents
                || _handledLossEpoch == feed.Epoch)
            {
                return false;
            }

            _handledLossEpoch = feed.Epoch;
            invalidation = _processor.DiscardUnconfirmed();
        }

        EmitSubtitleUpdate(invalidation.Value);
        return true;
    }

    public void EmitSubtitleUpdate(RealtimeSubtitleUpdate update)
    {
        RealtimeSubtitleUpdate stamped;
        lock (_sync)
        {
            stamped = update.Sequence == 0
                ? update with { Sequence = ++_subtitleSequence }
                : update;
        }

        _emitUpdate(stamped);
    }
}

File: RealtimeTranslator/Realtime/SessionSubtitlePipeline.swift
import Foundation

/// 字幕パイプラインの帳簿。`RealtimeSubtitleProcessor`（struct）、
/// `AudioLossTracker`（struct）、loss runToken を所有し、
/// aggregator / displayScheduler への描画を束ねる。@MainActor 閉じ込め。
@MainActor
final class SessionSubtitlePipeline {
    let aggregator: SubtitleAggregator
    let displayScheduler: SubtitleDisplayScheduler
    private var processor = RealtimeSubtitleProcessor()
    private var audioLossTracker = AudioLossTracker()
    private var handledLossRunToken: Int?

    private let postStopSubtitleRetentionNanoseconds: UInt64
    private let healthBookkeeper: SessionHealthBookkeeper
    private let dualClient: any DualRealtimeTranslationClienting

    /// セッション側の状態を読む provider。init 後に bind で結ぶ（self 捕捉のため）。
    private var activeFeedProvider: @MainActor () -> EventFeed? = { nil }
    private var lifecycleGenerationProvider: @MainActor () -> Int = { 0 }
    private var stateProvider: @MainActor () -> TranslationState = { .idle }
    private var publishUpdate: @MainActor (SubtitleSnapshot) -> Void = { _ in }

    init(
        aggregator: SubtitleAggregator,
        displayScheduler: SubtitleDisplayScheduler,
        postStopSubtitleRetentionNanoseconds: UInt64,
        healthBookkeeper: SessionHealthBookkeeper,
        dualClient: any DualRealtimeTranslationClienting
    ) {
        self.aggregator = aggregator
        self.displayScheduler = displayScheduler
        self.postStopSubtitleRetentionNanoseconds = postStopSubtitleRetentionNanoseconds
        self.healthBookkeeper = healthBookkeeper
        self.dualClient = dualClient
        self.displayScheduler.delegate = self
    }

    func bind(
        activeFeed: @escaping @MainActor () -> EventFeed?,
        lifecycleGeneration: @escaping @MainActor () -> Int,
        state: @escaping @MainActor () -> TranslationState,
        publishUpdate: @escaping @MainActor (SubtitleSnapshot) -> Void
    ) {
        activeFeedProvider = activeFeed
        lifecycleGenerationProvider = lifecycleGeneration
        stateProvider = state
        self.publishUpdate = publishUpdate
    }

    var audioLossMetrics: AudioLossMetrics {
        audioLossTracker.metrics
    }

    var activeLanguagePair: LanguagePair? {
        processor.activeLanguagePair
    }

    var isCurrentSegmentTainted: Bool {
        processor.isCurrentSegmentTainted
    }

    var hasSelectedTranslationTarget: Bool {
        processor.hasSelectedTranslationTarget
    }

    func beginEpoch(_ epoch: Int, pair: LanguagePair) {
        processor.beginEpoch(epoch, pair: pair)
    }

    func deactivateLanguagePair() {
        processor.deactivateLanguagePair()
    }

    func clearBoundaryCandidate() {
        processor.clearBoundaryCandidate()
    }

    func resetHandledLossRunToken() {
        handledLossRunToken = nil
    }

    func resetAudioLoss() {
        audioLossTracker.reset()
    }

    func observeAudio(
        generation: Int,
        sequence: Int,
        discardedMilliseconds: Int,
        queueWaitMilliseconds: Int,
        atMilliseconds: Int
    ) -> AudioLossObservation {
        audioLossTracker.observe(
            generation: generation,
            sequence: sequence,
            discardedMilliseconds: discardedMilliseconds,
            queueWaitMilliseconds: queueWaitMilliseconds,
            atMilliseconds: atMilliseconds
        )
    }

    func discardFailedSource(itemID: String?, eventID: String?) -> RealtimeSubtitleUpdate? {
        processor.discardFailedSource(itemID: itemID, eventID: eventID)
    }

    func discardUnconfirmed() -> RealtimeSubtitleUpdate {
        processor.discardUnconfirmed()
    }

    func tickProcessor(now: Date) -> RealtimeSubtitleUpdate? {
        processor.tick(now: now)
    }

    func takePendingUpdate() -> RealtimeSubtitleUpdate? {
        displayScheduler.takePendingUpdate()
    }

    func renderNow(_ update: RealtimeSubtitleUpdate) {
        displayScheduler.renderNow(update)
    }

    func discardPending() {
        displayScheduler.discardPending()
    }

    /// 正常停止の close drain で届いた字幕イベントを assembler へ取り込む。
    func ingestStopDrainEvents(_ events: [RealtimeTranslationStreamEvent], feed: EventFeed) {
        guard !feed.deliveryState.didLoseEvents else { return }
        for streamEvent in events {
            if case .error = streamEvent.event {
                continue
            }
            if case .inputTranscriptFailed(let itemID, let eventID, _, _) = streamEvent.event {
                guard streamEvent.epoch == feed.runToken else { continue }
                let invalidation = processor.discardFailedSource(itemID: itemID, eventID: eventID)
                feed.deliveryState.noteSourceFailureConsumed()
                if let invalidation {
                    displayScheduler.renderNow(invalidation)
                }
                continue
            }
            guard let result = processSubtitleEvent(streamEvent, now: Date(), isReplay: true) else {
                continue
            }
            for update in result.updates {
                displayScheduler.renderNow(update)
            }
        }
    }

    func schedulePostStopSubtitleClearIfNeeded() {
        cancelPostStopClear()
        guard !aggregator.snapshot().current.isEmpty else { return }
        let generation = lifecycleGenerationProvider()
        displayScheduler.schedulePostStopClear(
            afterNanoseconds: postStopSubtitleRetentionNanoseconds
        ) { [weak self] in
            guard let self else { return false }
            return self.lifecycleGenerationProvider() == generation && self.stateProvider() == .idle
        } onClear: { [weak self] in
            guard let self else { return }
            self.aggregator.reset()
            self.publishUpdate(self.aggregator.snapshot())
        }
    }

    func cancelPostStopClear() {
        displayScheduler.cancelPostStopClear()
    }

    /// 完全な原文+訳文ペアが assembler / aggregator に残っていれば idle 待ちを飛ばして確定する。
    /// 停止・再接続・致命エラーで epoch/buffer を捨てる直前に呼び、字幕記録の欠落を防ぐ。
    func flushPendingFinalizeIfNeeded() {
        if let feed = activeFeedProvider(),
            checkEventLoss(feed, generation: lifecycleGenerationProvider())
        {
            return
        }
        if let feed = activeFeedProvider(), feed.deliveryState.hasPendingSourceFailure {
            discardFailedSourceIfNeeded(feed)
            return
        }
        // スロットル中の live snapshot より assembler を正とする。
        displayScheduler.discardPending()

        if processor.isCurrentSegmentTainted {
            let invalidation = processor.discardUnconfirmed()
            displayScheduler.renderNow(invalidation)
            return
        }

        let flushAt = Date().addingTimeInterval(RealtimeSubtitleAssembler.idleFinalizeInterval)
        if let update = processor.tick(now: flushAt) {
            displayScheduler.renderNow(update)
            return
        }

        // assembler が空でも、live 経路 (canFinalize: false) の完全ペアが
        // aggregator に残っている場合がある — 字幕記録のため確定する。
        let before = aggregator.snapshot().current
        guard before.state != .finalized else { return }
        let snapshot = aggregator.forceFinalize()
        guard snapshot.current.state == .finalized else { return }
        publishUpdate(snapshot)
    }

    func discardFailedSourceIfNeeded(_ feed: EventFeed) {
        guard feed.deliveryState.hasPendingSourceFailure else { return }
        if let invalidation = processor.discardFailedSource(itemID: nil, eventID: nil) {
            displayScheduler.discardPending()
            displayScheduler.renderNow(invalidation)
        }
        while feed.deliveryState.hasPendingSourceFailure {
            feed.deliveryState.noteSourceFailureConsumed()
        }
    }

    func enqueueRender(_ update: RealtimeSubtitleUpdate) {
        if let feed = activeFeedProvider(),
            checkEventLoss(feed, generation: lifecycleGenerationProvider())
        {
            return
        }
        if update.isInvalidation {
            displayScheduler.discardPending()
            displayScheduler.renderNow(update)
            return
        }
        displayScheduler.enqueue(update)
    }

    func processSubtitleEvent(
        _ streamEvent: RealtimeTranslationStreamEvent,
        now: Date,
        isReplay: Bool = false
    ) -> RealtimeSubtitleProcessingResult? {
        if let feed = activeFeedProvider(),
            checkEventLoss(feed, generation: lifecycleGenerationProvider())
        {
            return nil
        }
        return processor.process(streamEvent, now: now, isReplay: isReplay)
    }

    func resetAudioRoutingForNextSegment() async {
        processor.resetRoutingForNextSegment()
        healthBookkeeper.setSelectedLane(nil)
        await dualClient.resetAudioRouting()
    }

    func handleAudioLoss() async {
        let invalidation = processor.markAudioLoss(now: Date())
        displayScheduler.discardPending()
        displayScheduler.renderNow(invalidation)
        if !processor.hasSelectedTranslationTarget {
            await dualClient.resetAudioRouting()
        }
    }

    /// scheduler からの描画要求を aggregator・delegate へ反映する。
    private func apply(_ update: RealtimeSubtitleUpdate) {
        if update.isInvalidation {
            let snapshot = aggregator.invalidateCurrent()
            publishUpdate(snapshot)
            return
        }

        if stateProvider() == .listening || stateProvider() == .reconnecting {
            aggregator.setStatusBanner(nil)
        }

        if update.shouldFinalize {
            let snapshot = aggregator.finalizePair(
                sourceText: update.sourceText,
                translatedText: update.translatedText,
                clearCurrent: true
            )
            publishUpdate(snapshot)
            return
        }

        let snapshot = aggregator.replaceCurrent(
            sourceText: update.sourceText,
            translatedText: update.translatedText,
            isTranslationCurrent: update.isTranslationCurrent,
            canFinalize: false
        )
        publishUpdate(snapshot)
    }

    func checkEventLoss(_ feed: EventFeed, generation _: Int) -> Bool {
        guard feed.deliveryState.didLoseEvents else { return false }
        if handledLossRunToken != feed.runToken {
            handledLossRunToken = feed.runToken
            let invalidation = processor.discardUnconfirmed()
            displayScheduler.discardPending()
            displayScheduler.renderNow(invalidation)
        }
        return true
    }

    func handleEventLoss(_ feed: EventFeed) {
        _ = checkEventLoss(feed, generation: lifecycleGenerationProvider())
    }
}

extension SessionSubtitlePipeline: SubtitleDisplaySchedulerDelegate {
    func subtitleDisplayScheduler(
        _ scheduler: SubtitleDisplayScheduler,
        requestsRenderOf update: RealtimeSubtitleUpdate
    ) {
        apply(update)
    }
}
